Fix #4648: Fix topic lesson differentiation between locked, completed and in-progress lessons

kkmurerwa opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fix #4648: When this PR is merged, it will fix the issue of locked, completed and in progress lessons not being differentiated by Talkback.

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

kkmurerwa committed:
fix: Fix #4648 where content for talkback in topic lessons was not differentiating between locked, completed and in-progress lessons

BenHenning commented on app/src/main/java/org/oppia/android/app/topic/lessons/ChapterSummaryViewModel.kt:
```diff
@@ -23,14 +24,22 @@ class ChapterSummaryViewModel(
+      }
+      ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES -> {
+        resourceHandler.getStringInLocaleWithWrapping(
+          R.string.chapter_prerequisite_title_label, (index).toString(), previousChapterTitle
```
```suggestion
          R.string.chapter_prerequisite_title_label, index.toString(), previousChapterTitle
```

kkmurerwa replied:
This has been modified with the latest changes.

BenHenning commented on app/src/main/java/org/oppia/android/app/topic/lessons/StorySummaryViewModel.kt:
```diff
@@ -77,6 +77,14 @@ class StorySummaryViewModel(
+            ephemeralStorySummary.chaptersList[index - 1].chapterSummary.title,
+            ephemeralStorySummary.chaptersList[index - 1].writtenTranslationContext
+          )
+        } else {
+          ""
+        },
```
Perhaps instead pass in null and add the null check to the 'when' condition for MISSING_PREREQUISITES? I think that will lead to a slightly cleaner & more robust implementation.

kkmurerwa replied:
I have added the else statement as suggested. It returns  a generic string that I also added to address the usecase where there is no previous chapter title.

BenHenning commented on app/src/main/java/org/oppia/android/app/topic/lessons/ChapterSummaryViewModel.kt:
```diff
@@ -23,14 +24,22 @@ class ChapterSummaryViewModel(
+          R.string.chapter_completed, (index + 1).toString(), chapterTitle
+        )
+      }
+      ChapterPlayState.NOT_PLAYABLE_MISSING_PREREQUISITES -> {
```
Please add a test for this follow-up change & add two screenshots to the PR description: one showing the new test failing when your fix is temporarily commented out/removed, and a second showing it passing. This is generally good practice when fixing targeted issues like this.

kkmurerwa replied:
I have added a test for the ideal situation (where there is a previous chapter title). I haven't been able to add a test for the other condition since I am unable to replicate it in test.

BenHenning replied:
Thanks! The other case is, theoretically, impossible. There is a way to force it to happen in tests, but I think it makes sense just to ignore it. We might be able to eliminate the codepath entirely with future refactors of the domain structures, but I think it's fine to leave it as-is untested.

BenHenning reviewed (changes requested):
Thanks @kkmurerwa! This seems like a really sensible solution. Had a few follow-up suggestions--PTAL.

kkmurerwa committed:
Merge branch 'develop' into differentiate-locked-completed-and-in-progress-lessons

kkmurerwa committed:
This PR will;
- Add a string for chapter locked state with no previous chapter title
- Add a default state for chapter locked with no previous chapter title
- Add a test for the new behaviour

kkmurerwa committed:
fix: Fix lints

kkmurerwa committed:
fix: Fix lints

kkmurerwa commented:
Screenshot(s) of the  failing test when the new code block is commented out;
![Screenshot from 2023-04-28 16-22-01](https://user-images.githubusercontent.com/18438114/235439084-6d4915c0-d395-4a74-86b5-e71be3a9d468.png)
The log produced by the failure
![Screenshot from 2023-04-28 16-22-45](https://user-images.githubusercontent.com/18438114/235439091-87cc107f-e862-43ca-a6ba-6e086fc0d7e9.png)

Screenshot of the passing test with the new code block uncommented;
![Screenshot from 2023-04-28 16-25-10](https://user-images.githubusercontent.com/18438114/235439220-0fc4f3fd-699e-4444-8759-cab25faaf5de.png)

kkmurerwa commented:
Hey @BenHenning. I have addressed your comments and added screenshots of the failing and passing test. You should be able to do another review. PTAL.

kkmurerwa commented:
Hey @BenHenning. I have temporarily reassigned myself the issue. There are a few more changes I am working on to improve accessibility.

kkmurerwa committed:
test: Add tests for the other two conditions that didn't previously exist

kkmurerwa commented:
Hey @BenHenning. Here are a few changes I have made with commit ID [8d497a2](https://github.com/oppia/oppia-android/pull/4958/commits/8d497a2c2e4be269cc3eed9fbe50a551162067f3).
- I noticed that there are no existing tests to check for the computeChapterPlayStateIconContentDescription that is displayed on the chapter_play_state_icon. I added a test for this that tests in the content description for progress chapters is being displayed correctly.
- In the case of a completed chapter topic, there is no chapter play state icon displayed. I modified the chapter_index text view to display the string from computeChapterPlayStateIconContentDescription instead for better Talkback feedback to visually impaired users.
- I added a test for the computeChapterPlayStateIconContentDescription in the case of the completed chapter topic state.


Here is a screenshot of the failing tests when I return a dummy string.
![Screenshot from 2023-05-02 18-31-56](https://user-images.githubusercontent.com/18438114/235716404-fbb7d18e-daf6-45f0-a721-e1d54462ceac.png)

Here is  a screenshot of the passing tests when the original logic is reinstated.
![Screenshot from 2023-05-02 18-35-12](https://user-images.githubusercontent.com/18438114/235716644-cf3599cc-5a0b-4815-a626-0635db7e02bb.png)

The tests on the CI failed for some reason but they all pass locallly. The failing test requires a rerun to pass.
![Screenshot from 2023-05-02 18-45-16](https://user-images.githubusercontent.com/18438114/235717461-1c49e18a-ebff-408c-a552-9cec719be585.png)

BenHenning commented:
Apologies @kkmurerwa will need to look at this tomorrow.

kkmurerwa commented:
No problem at all Ben.

BenHenning commented on app/src/main/java/org/oppia/android/app/topic/lessons/StorySummaryViewModel.kt:
```diff
@@ -77,6 +77,14 @@ class StorySummaryViewModel(
+            ephemeralStorySummary.chaptersList[index - 1].chapterSummary.title,
+            ephemeralStorySummary.chaptersList[index - 1].writtenTranslationContext
+          )
+        } else {
+          null
+        },
```
```suggestion
        } else null,
```

Small change to make it a bit more compact.

kkmurerwa replied:
This has been done too.

BenHenning commented on app/src/main/res/values/strings.xml:
```diff
@@ -91,6 +91,7 @@
   <string name="chapter_completed">Chapter %s with title %s is completed</string>
   <string name="chapter_in_progress">Chapter %s with title %s is in progress</string>
   <string name="chapter_prerequisite_title_label">Complete Chapter %s: %s to unlock this chapter.</string>
+  <string name="chapter_prerequisite_title_label_without_chapter_title">Complete the previous chapter to unlock this chapter</string>
```
```suggestion
  <string name="chapter_prerequisite_title_label_without_chapter_title">Complete the previous chapter to unlock this chapter.</string>
```

For consistency with the other label.

kkmurerwa replied:
I have made this change. It should reflect on the code now.

BenHenning reviewed (commented):
Thanks @kkmurerwa! PR LGTM, just had small nits--PTAL. Also, the branch will need to be brought up-to-date with the latest develop so that we can merge it.

kkmurerwa committed:
Merge branch 'develop' into differentiate-locked-completed-and-in-progress-lessons

kkmurerwa committed:
chore: Fix minor conciseness and consistency issues

kkmurerwa commented:
Hey @BenHenning. The issues mentioned have been resolved. The PR should now be ready for a merge unless there are some other issues.

kkmurerwa committed:
Merge branch 'develop' into differentiate-locked-completed-and-in-progress-lessons

BenHenning reviewed (approved):
Thanks @kkmurerwa! Apologies for the delayed review. This LGTM!
